File: TheaterSuggestions/Java/SeatsSuggestions/src/main/java/com/octo/technology/SeatsSuggestions/DeepModel/OfferSeatingPlacesNearerTheMiddleOfTheRow.java
package com.octo.technology.SeatsSuggestions.DeepModel;

import com.octo.technology.SeatsSuggestions.Row;
import com.octo.technology.SeatsSuggestions.Seat;
import com.octo.technology.SeatsSuggestions.SuggestionRequest;

import java.util.*;
import java.util.stream.Collectors;

public  class OfferSeatingPlacesNearerTheMiddleOfTheRow
{
    public static List<SeatWithDistance> buildSeatingPlaceCloserTheMiddleOfTheRow(Row row, SuggestionRequest suggestionRequest)
    {
        return makeSeatingPlacesWithDistance(row)
                .stream()
                .sorted(Comparator.comparingInt(SeatWithDistance::distanceFromTheMiddleOfTheRow))
                .filter(swd -> swd.seat().isAvailable())
                .filter(swd -> swd.seat().matchCategory(suggestionRequest.pricingCategory()))
                .collect(Collectors.toList());
    }

    public static List<SeatWithDistance> makeSeatingPlacesWithDistance(Row row)
    {
        int middleOfTheRow = row.seats().size() / 2;
        List<SeatWithDistance> seatingPlacesWithDistanceLeftSide =
                seatingPlacesWithDistanceForLeftSide(row, middleOfTheRow);
        List<SeatWithDistance> seatingPlacesWithDistanceRightSide =
                SeatingPlacesWithDistanceForRightSide(row, middleOfTheRow);

        seatingPlacesWithDistanceLeftSide.addAll(seatingPlacesWithDistanceRightSide);
        return seatingPlacesWithDistanceLeftSide;
    }

    private static List<SeatWithDistance> SeatingPlacesWithDistanceForRightSide(Row row, int middleOfTheRow)
    {
        return row.seats()
                .stream()
                .skip(middleOfTheRow)
                .map(s -> new SeatWithDistance(
                        s,
                        s.number() -  computeTheMiddleOfTheRowForRightSide(middleOfTheRow)
                ))
                .collect(Collectors.toList());
    }

    private static int computeTheMiddleOfTheRowForRightSide(int middleOfTheRow)
    {
        return (middleOfTheRow + 1);
    }

    private static List<SeatWithDistance> seatingPlacesWithDistanceForLeftSide(Row row, int middleOfTheRow)
    {
        return row.seats()
                .stream()
                .limit(middleOfTheRow)
                .map(s -> new SeatWithDistance(
                        s,
                        middleOfTheRow - ComputeTheMiddleOfTheRowForLeftSide(row, s)
                ))
                .collect(Collectors.toList());
    }

    private static int ComputeTheMiddleOfTheRowForLeftSide(Row row, Seat s)
    {
        return s.number() + Offset(row);
    }

    private static int Offset(Row row)
    {
        return RowIsEvent(row) ? 0 : -1;
    }

    private static boolean RowIsEvent(Row row)
    {
        return row.seats().size() % 2 == 0;
    }
}

File: TheaterSuggestions/CSharp/SeatsSuggestions.Domain/DeepModel/OfferSeatingPlacesNearerTheMiddleOfTheRow.cs
using System.Collections.Generic;
using System.Linq;

namespace SeatsSuggestions.Domain.DeepModel;

public static class OfferSeatingPlacesNearerTheMiddleOfTheRow
{
    public static List<SeatWithDistance> BuildSeatingPlaceCloserTheMiddleOfTheRow(Row row,
        SuggestionRequest suggestionRequest)
    {
        return MakeSeatingPlacesWithDistance(row)
            .OrderBy(swd => swd.DistanceFromTheMiddleOfTheRow)
            .Where(swd => swd.Seat.IsAvailable())
            .Where(swd => swd.Seat.MatchCategory(suggestionRequest.PricingCategory))
            .ToList();
    }

    public static IEnumerable<SeatWithDistance> MakeSeatingPlacesWithDistance(Row row)
    {
        var middleOfTheRow = row.Seats.Count / 2;
        var seatingPlacesWithDistanceLeftSide =
            SeatingPlacesWithDistanceForLeftSide(row, middleOfTheRow);
        var seatingPlacesWithDistanceRightSide =
            SeatingPlacesWithDistanceForRightSide(row, middleOfTheRow);
        return seatingPlacesWithDistanceLeftSide.Concat(seatingPlacesWithDistanceRightSide);
    }

    private static IEnumerable<SeatWithDistance> SeatingPlacesWithDistanceForRightSide(Row row, int middleOfTheRow)
    {
        return row.Seats
            .Skip(middleOfTheRow)
            .Select(s => new SeatWithDistance(
                s,
                (int)(s.Number - ComputeTheMiddleOfTheRowForRightSide(middleOfTheRow))
            ));
    }

    private static int ComputeTheMiddleOfTheRowForRightSide(int middleOfTheRow)
    {
        return middleOfTheRow + 1;
    }

    private static IEnumerable<SeatWithDistance> SeatingPlacesWithDistanceForLeftSide(Row row, int middleOfTheRow)
    {
        return row.Seats
            .Take(middleOfTheRow)
            .Select(s => new SeatWithDistance(
                s,
                middleOfTheRow - ComputeTheMiddleOfTheRowForLeftSide(row, s)
            ));
    }

    private static int ComputeTheMiddleOfTheRowForLeftSide(Row row, Seat s)
    {
        return (int)(s.Number + Offset(row));
    }

    private static int Offset(Row row)
    {
        return RowIsEvent(row) ? 0 : -1;
    }

    private static bool RowIsEvent(Row row)
    {
        return row.Seats.Count % 2 == 0;
    }
}